compiler/src/peephole.rs

```diff
@@ -0,0 +1,182 @@
+                    value: bytecode::Constant::$name {$($field)*},
+                }
+            };
+            ($name:ident, $($value:tt)*) => {
+                lc!($name { value: $($value)* })
+            };
+        }
+        macro_rules! emitconst {
+            ([$($metas:expr),*], $($arg:tt)*) => {
+                self.emit(
+                    lc!($($arg)*),
+                    InstructionMetadata::from(vec![$($metas),*]),
+                )
+            };
+        }
+        match instruction {
+            Instruction::BinaryOperation { op, inplace } => {
+                let (rhs, rhs_meta) = self.pop();
+                let (lhs, lhs_meta) = self.pop();
+                macro_rules! op {
+                    ($op:ident) => {
+                        bytecode::BinaryOperator::$op
+                    };
+                }
+                match (op, lhs, rhs) {
+                    (op!(Add), lc!(Integer, lhs), lc!(Integer, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Integer, lhs + rhs)
+                    }
+                    (op!(Subtract), lc!(Integer, lhs), lc!(Integer, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Integer, lhs - rhs)
+                    }
+                    (op!(Add), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs + rhs)
+                    }
+                    (op!(Subtract), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs - rhs)
+                    }
+                    (op!(Multiply), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs * rhs)
+                    }
+                    (op!(Divide), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs / rhs)
+                    }
+                    (op!(Power), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs.powf(rhs))
+                    }
+                    (op!(Add), lc!(String, mut lhs), lc!(String, rhs)) => {
+                        lhs.push_str(&rhs);
+                        emitconst!([lhs_meta, rhs_meta], String, lhs);
+                    }
+                    (op, lhs, rhs) => {
+                        self.emit(lhs, lhs_meta);
+                        self.emit(rhs, rhs_meta);
+                        self.emit(Instruction::BinaryOperation { op, inplace }, meta);
+                    }
+                }
+            }
+            Instruction::UnpackSequence { size } => {
+                let (arg, arg_meta) = self.pop();
+                match arg {
+                    Instruction::BuildTuple {
+                        size: tup_size,
+                        unpack,
+                    } if !unpack && tup_size == size => {
+                        self.emit(
```

Comment:
Does this not cause infinite recursion? Since `emit` calls `optimize` again?

Reply:
It doesn't, a method declared directly on a type takes precedence over a method with the same name that's part of a trait implemented for that type. I can see how this might be confusing; I'll rename the `PeepholeOptimizer::emit` method.